presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -319,7 +319,7 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
             }
 
             // analyze the query that creates the data
-            Scope queryScope = process(insert.getQuery(), scope);
```

Comment:
how `insert` can have an outer scope?

Reply:
It's a synthetic resolution scope for the "INSERT" statement that doesn't bind any variables. The SELECT query associated with the INSERT will see it as an outer parent scope and can reason about whether ORDER BY has any effect in that context.

Reply:
So you are going to mark `ORDER BY` with any outer scope set during semantic analysis and prune it during the planning?

To me it is a bit hacky because we don't set an outer scope for `SELECT` statement but we do for `INSERT` just because `ORDER BY` can be optimized. To me it should be handled entirely during the planing (mostly in optimizer).

Reply:
No, the way it works is that an order by for a query that’s in a nested scope (I.e, it’s not top-level, semantically speaking) is redundant and can be elided.

> To me it is a bit hacky because we don't set an outer scope for SELECT statement

We actually do. The call to “analyze” on a query object establishes a scope.